pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/driver.go

```diff
@@ -18,205 +18,193 @@ package tinkerbell
+		log.Infow("server is not allowed to be provisioned; either hardware allowPXE and allowWorkflow is set tp false", "hardware", hardware.Name)
+		return nil, nil
 	}
 
-	// cfg.SecretName has the same name as the machine name
-	workflowTemplate, err := d.templateClient.Get(ctx, "", cfg.SecretName)
+	// Create template if it doesn't exist
+	err = d.TemplateClient.CreateTemplate(ctx, hardware, d.HardwareRef.Namespace, d.OSImageURL)
 	if err != nil {
-		if resourceNotFoundErr(err) {
-			tmpl := createTemplate(d.TinkServerAddress, d.ImageRepoAddress, cfg)
-			payload, err := yaml.Marshal(tmpl)
-			if err != nil {
-				return nil, fmt.Errorf("failed marshalling workflow template: %w", err)
-			}
-
-			workflowTemplate = &tinktmpl.WorkflowTemplate{
-				Name: tmpl.Name,
-				Id:   tmpl.ID,
-				Data: string(payload),
-			}
-
-			if err := d.templateClient.Create(ctx, workflowTemplate); err != nil {
-				return nil, fmt.Errorf("failed to create workflow template: %w", err)
-			}
-		}
+		return nil, err
+	}
+
+	// Create Workflow to match the template and server
+	server := tinktypes.Hardware{Hardware: hardware}
+	if err = d.WorkflowClient.CreateWorkflow(ctx, userdata, server.Name, client.ProvisionWorkerNodeTemplate, server); err != nil {
+		return nil, err
 	}
 
-	if _, err := d.workflowClient.Create(ctx, workflowTemplate.Id, hw.GetID()); err != nil {
-		return nil, fmt.Errorf("failed to provision server id %s running template id %s: %w", workflowTemplate.Id, hw.GetID(), err)
+	// Set the HardwareID with machine UID. The hardware object is claimed by the machine.
+	if err = d.HardwareClient.SetHardwareID(ctx, hardware, string(meta.UID)); err != nil {
+		return nil, err
 	}
 
-	return &hw, nil
+	return &server, nil
 }
 
-func (d *driver) Validate(hwSpec runtime.RawExtension) error {
-	hw := HardwareSpec{}
-	if err := json.Unmarshal(hwSpec.Raw, &hw); err != nil {
-		return fmt.Errorf("failed to unmarshal tinkerbell hardware spec: %w", err)
-	}
+func (d *driver) Validate(_ runtime.RawExtension) error {
+	return nil
+}
 
-	if hw.Hardware.Hardware == nil {
-		return fmt.Errorf("tinkerbell hardware data can not be empty")
+func (d *driver) DeprovisionServer(ctx context.Context) error {
+	// Get the hardware object from tinkerbell cluster
+	targetHardware, err := d.HardwareClient.GetHardware(ctx, d.HardwareRef)
+	if err != nil {
+		return err
 	}
 
-	if hw.Hardware.Network == nil {
-		return fmt.Errorf("tinkerbell hardware network configs can not be empty")
+	// Delete the associated Workflow.
+	workflowName := targetHardware.Name + "-workflow" // Assuming workflow names are derived from hardware names
+	if err := d.WorkflowClient.DeleteWorkflow(ctx, workflowName, targetHardware.Namespace); err != nil {
+		return fmt.Errorf("failed to delete workflow %s: %w", workflowName, err)
 	}
 
-	if hw.Hardware.Metadata == "" {
-		return fmt.Errorf("tinkerbell hardware metadata can not be empty")
+	// Reset the hardware ID in the tinkerbell cluster.
```

Comment:
ID _and state_, so it's more obvious that from this point on, the machine is also not Staged anymore.